game/procurement.py

```diff
@@ -116,52 +128,236 @@ def repair_runways(self, budget: float) -> float:
+
+            unit = self.buy_ground_unit(
+                ground_unit_budget, most_needed_type, cp_with_fewest_units
+            )
             if unit is None:
-                # Can't afford any more units.
                 break
 
-            budget -= db.PRICES[unit]
-            cp.pending_unit_deliveries.order({unit: 1})
+            unit_price = db.PRICES[unit]
+            ground_unit_budget -= unit_price
+            cp_with_fewest_units.pending_unit_deliveries.order({unit: 1})
 
-        return budget
+        return ground_unit_budget
+
+    def buy_ground_unit(
+        self, budget: float, vehicle_type: List[Type[VehicleType]], cp: ControlPoint
+    ) -> Optional[Type[VehicleType]]:
+        unit = self.random_affordable_ground_unit(budget, vehicle_type)
+
+        if unit is None:
+            # Can't afford any more units.
+            return None
+
+        return unit
+
+    def calculate_vehicle_investment_ratio(
+        self, cp: ControlPoint
+    ) -> List[Tuple[List, float]]:
+        prioritylist: List[Tuple[List, float]] = []
+
+        tank_costs = self.calculate_investment_of_vehicleType(cp, TYPE_TANKS)
+        atgm_costs = self.calculate_investment_of_vehicleType(cp, TYPE_ATGM)
+        ifv_costs = self.calculate_investment_of_vehicleType(cp, TYPE_IFV)
+        apc_costs = self.calculate_investment_of_vehicleType(cp, TYPE_APC)
+        artillery_costs = self.calculate_investment_of_vehicleType(cp, TYPE_ARTILLERY)
+        shorad_costs = self.calculate_investment_of_vehicleType(cp, TYPE_SHORAD)
+
+        costs_of_all_vehicles = (
+            tank_costs
+            + atgm_costs
+            + ifv_costs
+            + apc_costs
+            + artillery_costs
+            + shorad_costs
+        )
+
+        faction_has_tank_access = False
+        faction_has_atgm_access = False
+        faction_has_ifv_access = False
+        faction_has_apc_access = False
+        faction_has_artillery_access = False
+        faction_has_shorad_access = False
+
+        for unit in self.faction.frontline_units:
+            if faction_has_tank_access is False and unit in TYPE_TANKS:
+                faction_has_tank_access = True
+            if faction_has_atgm_access is False and unit in TYPE_ATGM:
+                faction_has_atgm_access = True
+            if faction_has_ifv_access is False and unit in TYPE_IFV:
+                faction_has_ifv_access = True
+            if faction_has_apc_access is False and unit in TYPE_APC:
+                faction_has_apc_access = True
+            if faction_has_shorad_access is False and unit in TYPE_SHORAD:
+                faction_has_shorad_access = True
+
+        for unit in self.faction.artillery_units:
+            if faction_has_artillery_access is False and unit in TYPE_ARTILLERY:
+                faction_has_artillery_access = True
+                break
+
+        if faction_has_tank_access:
+            if costs_of_all_vehicles is 0:
```

Comment:
`is` and `==` are not the same. `is` is reference equality.

```python
>>> x = 1000
>>> y = 1000
>>> x is y
False
```

Apparently with ints this can be extremely misleading (https://stackoverflow.com/a/306353/632035), and because of that it'll work with 0, but that's an implementation detail we should not depend on :)